common/src/main/java/io/netty/util/internal/PlatformDependent.java

```diff
@@ -278,6 +273,61 @@ public static long byteArrayBaseOffset() {
         return BYTE_ARRAY_BASE_OFFSET;
     }
 
+    private static Data getData() {
+        if (DATA == null) {
+            initialize();
+        }
+        return DATA;
+    }
+
+    private static final Object lock = new Object();
+    private static void initialize() {
+        synchronized (lock) {
+            if (DATA == null) {
+                final boolean useDirectBufferNoCleaner;
+                final AtomicLong directMemoryCounter;
+
+                final long estimateMaxDirectMemory = estimateMaxDirectMemory();
+
+                // Here is how the system property is used:
+                //
+                // * <  0  - Don't use cleaner, and inherit max direct memory from java. In this case the
+                //           "practical max direct memory" would be 2 * max memory as defined by the JDK.
+                // * == 0  - Use cleaner, Netty will not enforce max memory, and instead will defer to JDK.
+                // * >  0  - Don't use cleaner. This will limit Netty's total direct memory
+                //           (note: that JDK's direct memory limit is independent of this).
+                long maxDirectMemory = MaxDirectMemorySetting.get();
+                if (maxDirectMemory >= 0) {
+                    logger.debug("max direct memory was defined using a setter method from the class '{}'",
+                        MaxDirectMemorySetting.class.getName());
+                }
+
+                if (maxDirectMemory < 0) {
+                    maxDirectMemory = estimateMaxDirectMemory;
+                }
+
+                if (maxDirectMemory == 0 || !hasUnsafe() || !PlatformDependent0.hasDirectBufferNoCleanerConstructor()) {
+                    useDirectBufferNoCleaner = false;
+                    directMemoryCounter = null;
+                } else {
+                    useDirectBufferNoCleaner = true;
+                    if (maxDirectMemory < 0) {
+                        if (maxDirectMemory <= 0) {
```

Comment:
I think this needs to be kept in line with https://github.com/netty/netty/blob/937bb67360a5b4ea0a6e11bbe24c86c76a4d6620/common/src/main/java/io/netty/util/internal/PlatformDependent.java#L167-L169. Originally, the `maxDirectMemory` variable used to be updated between the two ifs of the snippet below. That logic should remain as it was originally: having the if block `maxDirectMemory < 0` outside changes it.

@manseaume can you check the proposed change as you last worked on that ? i.e., https://github.com/sirensolutions/netty/pull/26

```suggestion
                if (maxDirectMemory == 0 || !hasUnsafe() || !PlatformDependent0.hasDirectBufferNoCleanerConstructor()) {
                    useDirectBufferNoCleaner = false;
                    directMemoryCounter = null;
                } else {
                    useDirectBufferNoCleaner = true;
                    if (maxDirectMemory < 0) {
                        maxDirectMemory = estimateMaxDirectMemory;
                        if (maxDirectMemory <= 0) {
```

Reply:
Agreed, it's better for future upgrades to stay close to original netty code.

Reply:
Done https://github.com/sirensolutions/netty/pull/29/commits/02ef1363cc952aaffb29ed3a4384f50aa2909edb

Reply:
@fabiocorneti the changes proposed above reverts the changes you made here https://github.com/sirensolutions/netty/pull/17/commits/987104864a04c5ad0bb59eda3f56872a8c1f64eb and https://github.com/sirensolutions/netty/pull/17/commits/4a45f0c269b3d12b51d84116b616ad12c9a787ca# was there a particular issue for this change or we could revert it back to the original netty design as proposed above?

Reply:
Found the reason in the discussion here https://github.com/sirensolutions/netty/pull/1#discussion_r433911346  and reverted to the changes that existed https://github.com/sirensolutions/netty/pull/29/commits/f02494a215cf20904629142f81e49d0150bdd683

Reply:
Thanks @GeorgeAp ! That's very useful